### PR TITLE
Support to get attribute of date/time/datetime.

### PR DESCRIPTION
**Pull request by Shylock-Hg**
We shouldn't process the bad null different with null, they are all null in computation.

### CODE REVIEW

File: src/common/expression/AttributeExpression.cpp [hunk 1]
```diff
@@ -18,42 +19,54 @@ const Value& AttributeExpression::eval(ExpressionContext &ctx) {
...
-        if (rvalue.getStr() == kVid) {
-            result_ = lvalue.getVertex().vid;
-            return result_;
-        }
-        for (auto &tag : lvalue.getVertex().tags) {
-            auto iter = tag.props.find(rvalue.getStr());
-            if (iter != tag.props.end()) {
-                return iter->second;
+    switch (lvalue.type()) {
+        case Value::Type::MAP       :
```
Comment (laura-ding) [inline src/common/expression/AttributeExpression.cpp:23]: Delete the space between `:`
under ditto

File: src/common/time/TimeUtils.h [hunk 2]
```diff
@@ -92,6 +92,26 @@ class TimeUtils {
...
+        } else if (prop == "hour") {
+            return static_cast<int>(dt.hour);
+        } else if (prop == "minute") {
+            return static_cast<int>(dt.minute);
+        } else if (prop == "second") {
+            return static_cast<int>(dt.sec);
+        } else if (prop == "microsecond") {
+            return static_cast<int>(dt.microsec);
+        } else {
+            return Value::kNullValue;
```
Comment (laura-ding) [inline src/common/time/TimeUtils.h:111]: kNullUnknownProp

File: src/common/time/TimeUtils.h [hunk 3]
```diff
@@ -129,6 +149,19 @@ class TimeUtils {
...
 
+    static Value getDateAttr(const Date &d, const std::string &prop) {
+        if (prop == "year") {
+            return d.year;
+        } else if (prop == "month") {
+            return d.month;
+        } else if (prop == "day") {
+            return d.day;
+        } else {
+            return Value::kNullValue;
```
Comment (laura-ding) [inline src/common/time/TimeUtils.h:160]: kNullUnknownProp

File: src/common/time/TimeUtils.h [hunk 4]
```diff
@@ -171,6 +204,21 @@ class TimeUtils {
...
+        if (prop == "hour") {
+            return t.hour;
+        } else if (prop == "minute") {
+            return t.minute;
+        } else if (prop == "second") {
+            return t.sec;
+        } else if (prop == "microsecond") {
+            return t.microsec;
+        } else {
+            return Value::kNullValue;
```
Comment (laura-ding) [inline src/common/time/TimeUtils.h:217]: kNullUnknownProp

File: src/common/expression/AttributeExpression.cpp (hunk 1 above)
Comment (Shylock-Hg) [reply src/common/expression/AttributeExpression.cpp]: Fixed.

File: src/common/time/TimeUtils.h (hunk 2 above)
Comment (Shylock-Hg) [reply src/common/time/TimeUtils.h:111]: It's the previous return value.

**Review by laura-ding — APPROVED** [review]

**Review by yixinglu — DISMISSED** [review]

File: src/common/expression/AttributeExpression.cpp [hunk 5]
```diff
@@ -18,42 +19,54 @@ const Value& AttributeExpression::eval(ExpressionContext &ctx) {
...
+            }
+            auto iter = lvalue.getEdge().props.find(rvalue.getStr());
+            if (iter == lvalue.getEdge().props.end()) {
+                return Value::kNullUnknownProp;
+            }
+            return iter->second;
         }
-        return iter->second;
+        case Value::Type::DATE:
+            result_ = time::TimeUtils::getDateAttr(lvalue.getDate(), rvalue.getStr());
```
Comment (yixinglu) [inline src/common/expression/AttributeExpression.cpp:59]: ::tolower(rvale.getStr())

Comment (Shylock-Hg) [reply src/common/expression/AttributeExpression.cpp:59]: Fixed.

File: src/common/time/TimeUtils.h [hunk 6]
```diff
@@ -134,6 +158,20 @@ class TimeUtils {
         return TimeConversion::unixSecondsToDate(unixTime);
     }
 
+    static Value getDateAttr(const Date &d, const std::string &prop) {
+        auto lowerProp = boost::algorithm::to_lower_copy(prop);
+        if (lowerProp == "year") {
+            return d.year;
```
Comment (yixinglu) [inline src/common/time/TimeUtils.h:164]: Don't you need to cast the type here ?

File: src/common/time/TimeUtils.h [hunk 7]
```diff
@@ -176,6 +214,22 @@ class TimeUtils {
         return t;
     }
 
+    static Value getTimeAttr(const Time &t, const std::string &prop) {
+        auto lowerProp = boost::algorithm::to_lower_copy(prop);
+        if (lowerProp == "hour") {
+            return t.hour;
```
Comment (yixinglu) [inline src/common/time/TimeUtils.h:220]: ditto

File: src/common/time/TimeUtils.h (hunk 6 above)
Comment (Shylock-Hg) [reply src/common/time/TimeUtils.h:164]: Why need cast?

Comment (yixinglu) [reply src/common/time/TimeUtils.h:164]: it's not same as above getDateTimeAttr function.

Comment (Shylock-Hg) [reply src/common/time/TimeUtils.h:164]: The datetime is bitfield, so need cast to fit the Value constructor.

**Review by laura-ding — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**